TLM/TLM/Manager/Impl/LaneConnectionManagerData/ConnectionData.cs

```diff
@@ -0,0 +1,159 @@
+            for(int i = 0; i < n; ++i) {
+                if(targets[i].LaneId == targetLaneId) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Adds a connection from source to target lane at the give node
+        /// </summary>
+        /// <param name="sourceStartNode">start node for the segment of the source lane</param>
+        internal void ConnectTo(uint sourceLaneId, uint targetLaneId, bool sourceStartNode) {
+            ref var targets = ref GetConnections(sourceLaneId, sourceStartNode);
+            int n = targets?.Length ?? 0;
+            for(int i = 0; i < n; ++i) {
+                if(targets[i].LaneId == targetLaneId) {
+                    return; // existing connection
+                }
+            }
+
+            targets = targets.AppendOrCreate(new LaneConnectionData(targetLaneId));
+        }
+
+        /// <summary>
+        /// Adds a connection from source to target lane at the give node
+        /// </summary>
+        internal void ConnectTo(uint sourceLaneId, uint targetLaneId, ushort nodeId) =>
+            ConnectTo(sourceLaneId, targetLaneId, sourceLaneId.ToLane().IsStartNode(nodeId));
+
+        /// <summary>removes the connection from source to target lane at the given node</summary>
+        internal bool Disconnect(uint sourceLaneId, uint targetLaneId, ushort nodeId) =>
+            Disconnect(sourceLaneId, targetLaneId, sourceLaneId.ToLane().IsStartNode(nodeId));
+
+        /// <summary>removes the connection from source to target lane at the given node</summary>
+        /// <param name="sourceStartNode">start node for the segment of the source lane</param>
+        internal bool Disconnect(uint sourceLaneId, uint targetLaneId, bool sourceStartNode) {
+            ref var targets = ref GetConnections(sourceLaneId, sourceStartNode);
+            if(targets == null)
+                return false;
+
+            bool ret = false;
+            var newConnections = new List<LaneConnectionData>(targets.Length);
+            for(int i = 0; i < targets.Length; ++i) {
+                if(targets[i].LaneId != targetLaneId) {
+                    newConnections.Add(targets[i]);
+                } else {
+                    ret = true;
+                }
+            }
+
+            if(newConnections.Count == 0)
+                targets = null;
+            else
+                targets = newConnections.ToArray();
+
+            return ret;
+        }
+
+        /// <summary>
+        /// removes all connections from and to the given lane.
+        /// </summary>
+        internal void RemoveConnections(uint laneId) {
+            connections_[laneId].StartConnections = connections_[laneId].EndConnections = null;
+            for(uint laneId2 = 1; laneId2 < connections_.Length; ++laneId2) {
+                Disconnect(sourceLaneId: laneId2, targetLaneId: laneId, sourceStartNode: true);
```

Comment:
ton is a relative word. its minuscule in comparison to what CS does in background 😄 
It only happens when user removes a connection so its fine. its not performance critical.
Removing connections FROM the lane is easy but I do not see a better way of removing connections TO the lane.

Reply:
Are you sure? You are looping through 260k lanes and each `Disconnect` creates a more or less useless `List<>` inside if there is at least one connection. It's not a problem if you have 200 connections in total, but users spam stay-on-lane everywhere so we are talking about thousands of useless allocations just to find if there is a lane and remove it... 
Maybe use static `FastList<>` as a helper and clear after use? It's an array that only grows unless you explicitly call `Trim`.

Reply:
this only happens when you delete a segment. 
I don't think you can have that many connections at a segment.
i guess if user deletes a big chunk of the city with MoveIt at once then maybe we can feel the effect? my PC is too fast for me to notice these things so that is also something to consider.